File: scripts/v2/remediation/egress_dlp.py
"""egress_dlp.py — Python parity of web/lib/egress-dlp.ts (ADR-040 §2 exfiltration defense).

Re-applied inside the executor (never trust the web layer). Pure (no I/O). Same patterns + entropy-gated
long-blob heuristic + size cap + channel allowlist as the TS version — the executor is the LAST hop, so
its redaction is the final guarantee before a message leaves to an external SaaS. Best-effort by design;
the human 4-eyes review of the dry-run preview is the real backstop.
"""
import re

_SIZE_CAP = 3000
_TRUNC = "…[truncated]"
_VARIETY_MIN = 12

_PATTERNS = [
    (re.compile(r"\b(?:AKIA|ASIA)[A-Z0-9]{16}\b"), "aws-key"),
    (re.compile(r"arn:aws:[a-z0-9-]*:[a-z0-9-]*:\d{0,12}:[^\s\"']+", re.I), "arn"),
    (re.compile(r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+"), "jwt"),
    (re.compile(r"\bBearer\s+[A-Za-z0-9._-]+", re.I), "bearer"),
    (re.compile(r"aws_secret[a-z_]*\s*[=:]\s*\S+", re.I), "aws-secret"),
    (re.compile(
        r"\b(?:10\.\d{1,3}\.\d{1,3}\.\d{1,3}|172\.(?:1[6-9]|2\d|3[01])\.\d{1,3}\.\d{1,3}"
        r"|192\.168\.\d{1,3}\.\d{1,3}|169\.254\.\d{1,3}\.\d{1,3})\b"), "private-ip"),
]
_BLOB = re.compile(r"[A-Za-z0-9+/]{40,}={0,2}|[0-9a-fA-F]{40,}")


def _redact_str(s, cats):
    for rx, cat in _PATTERNS:
        s, n = rx.subn(f"[REDACTED:{cat}]", s)
        if n:
            cats.add(cat)

    def _blob(m):
        t = m.group(0)
        if len(set(t)) < _VARIETY_MIN:   # low-entropy repeat → leave for the size cap (not a secret)
            return t
        cats.add("blob")
        return "[REDACTED:blob]"

    s = _BLOB.sub(_blob, s)
    if len(s) > _SIZE_CAP:
        cats.add("size-cap")
        s = s[: _SIZE_CAP - len(_TRUNC)] + _TRUNC
    return s


def _redact_val(v, cats):
    if isinstance(v, str):
        return _redact_str(v, cats)
    if isinstance(v, list):
        return [_redact_val(x, cats) for x in v]
    if isinstance(v, dict):
        return {k: _redact_val(val, cats) for k, val in v.items()}
    return v


def redact_egress(payload):
    """Redact secrets/internal data from EVERY string field (recurses). Returns (payload, sorted categories)."""
    cats = set()
    out = _redact_val(payload, cats)
    return out, sorted(cats)


class ChannelNotAllowed(Exception):
    pass


def assert_channel_allowed(channel, allowlist):
    """Raise unless `channel` is in `allowlist`. Empty/non-list allowlist = deny-all (fail-closed)."""
    if not isinstance(allowlist, list) or len(allowlist) == 0:
        raise ChannelNotAllowed("channel allowlist is empty (deny-all, fail-closed)")
    if channel not in allowlist:
        raise ChannelNotAllowed("channel %s is not in the allowlist" % channel)

File: web/lib/egress-dlp.ts
// web/lib/egress-dlp.ts
// ADR-040 §2 exfiltration defense (RW-slice T2) — redact secrets/internal data from an outbound
// (egress-WRITE) payload before it can leave to an external SaaS, and enforce a destination
// (channel) allowlist. HONEST LIMIT: regex DLP is best-effort — the mandatory human 4-eyes review of
// the dry-run preview is the real backstop; this catches the obvious + a long-opaque-blob heuristic
// for encoded secrets, the allowlist bounds destinations, and the agent has no raw-secret-read tool.
// Applied at the web action gate AND re-applied in the Python executor (never trust the upstream).

const SIZE_CAP = 3000;
const TRUNC_MARKER = '…[truncated]';

// Specific secret/internal-data patterns → [REDACTED:<cat>].
const PATTERNS: { re: RegExp; cat: string }[] = [
  { re: /\b(?:AKIA|ASIA)[A-Z0-9]{16}\b/g, cat: 'aws-key' },
  { re: /arn:aws:[a-z0-9-]*:[a-z0-9-]*:\d{0,12}:[^\s"']+/gi, cat: 'arn' },
  { re: /\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/g, cat: 'jwt' },
  { re: /\bBearer\s+[A-Za-z0-9._-]+/gi, cat: 'bearer' },
  { re: /aws_secret[a-z_]*\s*[=:]\s*\S+/gi, cat: 'aws-secret' },
  {
    re: /\b(?:10\.\d{1,3}\.\d{1,3}\.\d{1,3}|172\.(?:1[6-9]|2\d|3[01])\.\d{1,3}\.\d{1,3}|192\.168\.\d{1,3}\.\d{1,3}|169\.254\.\d{1,3}\.\d{1,3})\b/g,
    cat: 'private-ip',
  },
];

// Long opaque base64/hex blob heuristic — catches RAW dumps / long encoded payloads. Entropy-gated:
// only a HIGH-VARIETY run is masked; a low-variety run (e.g. 'xxxx…') falls through to the size cap.
// HONEST LIMIT (P4 gate): this does NOT catch a SHORT encoded secret — a bare 20-char AWS key base64s
// to ~28 chars, below the 40-char floor, so it is NOT redacted. The regex is best-effort; the
// exfiltration backstop is the mandatory human 4-eyes review of the dry-run preview (ADR-040 §2), not
// this heuristic. (Lowering the floor would false-positive on legit tokens; the design relies on the
// human gate + propose-only + no raw-secret-read tool, not on perfect regex.)
const BLOB_RE = /[A-Za-z0-9+/]{40,}={0,2}|[0-9a-fA-F]{40,}/g;
const VARIETY_MIN = 12;

function redactString(input: string, cats: Set<string>): string {
  let s = input;
  for (const { re, cat } of PATTERNS) {
    s = s.replace(re, () => { cats.add(cat); return `[REDACTED:${cat}]`; });
  }
  s = s.replace(BLOB_RE, (m) => {
    if (new Set(m).size < VARIETY_MIN) return m; // low-entropy (repeat) → leave for the size cap
    cats.add('blob');
    return '[REDACTED:blob]';
  });
  if (s.length > SIZE_CAP) {
    cats.add('size-cap');
    s = s.slice(0, SIZE_CAP - TRUNC_MARKER.length) + TRUNC_MARKER;
  }
  return s;
}

function redactValue(v: unknown, cats: Set<string>): unknown {
  if (typeof v === 'string') return redactString(v, cats);
  if (Array.isArray(v)) return v.map((x) => redactValue(x, cats));
  if (v && typeof v === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, val] of Object.entries(v as Record<string, unknown>)) out[k] = redactValue(val, cats);
    return out;
  }
  return v;
}

/** Redact secrets/internal data from EVERY string field of an egress-write payload (recurses into
 *  nested blocks/attachments). Returns the redacted payload + the deduped list of redaction categories
 *  (NOT the secret values). Idempotent on already-clean text. */
export function redactEgress<T>(payload: T): { payload: T; redactions: string[] } {
  const cats = new Set<string>();
  const out = redactValue(payload, cats) as T;
  return { payload: out, redactions: [...cats] };
}

/** Throw unless `channel` is in `allowlist`. An empty/missing allowlist is DENY-ALL (fail-closed) —
 *  the destination must be explicitly admin-allowlisted on the integration (egress-write source_allowlist). */
export function assertChannelAllowed(channel: string, allowlist: string[]): void {
  if (!Array.isArray(allowlist) || allowlist.length === 0) {
    throw new Error('egress-dlp: channel allowlist is empty (deny-all, fail-closed)');
  }
  if (!allowlist.includes(channel)) {
    throw new Error(`egress-dlp: channel ${channel} is not in the allowlist`);
  }
}
